File: ios/EscapePhone/Core/Services/PlatformServices.swift
import Foundation
import UIKit

enum HapticEvent { case success, error, selection, digitDiscovered }
protocol HapticProvider { func play(_ event: HapticEvent) }
struct PlatformHapticProvider: HapticProvider {
    func play(_ event: HapticEvent) {
        switch event {
        case .success: UINotificationFeedbackGenerator().notificationOccurred(.success)
        case .error: UINotificationFeedbackGenerator().notificationOccurred(.error)
        case .selection: UISelectionFeedbackGenerator().selectionChanged()
        case .digitDiscovered: UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        }
    }
}
struct NoOpHapticProvider: HapticProvider { func play(_ event: HapticEvent) {} }

enum AdResult: Equatable { case rewarded, skipped, unavailable, failed(String) }
protocol AdGateway { func showRewardedAd() async -> AdResult }
struct NoOpAdGateway: AdGateway { func showRewardedAd() async -> AdResult { .unavailable } }
final class FakeAdGateway: AdGateway {
    var result: AdResult = .rewarded
    init(result: AdResult = .rewarded) { self.result = result }
    func showRewardedAd() async -> AdResult { try? await Task.sleep(for: .milliseconds(350)); return result }
}

enum DeviceCommand: Equatable { case ping, arm, unlock, reset }
enum PuzzleDeviceState: Equatable { case disconnected, connecting, connected, error(String) }
protocol PuzzleDeviceConnector: AnyObject {
    var state: PuzzleDeviceState { get }
    func connect(sessionId: String) async
    func send(_ command: DeviceCommand) async throws
    func disconnect() async
}
final class NoOpPuzzleDeviceConnector: PuzzleDeviceConnector {
    private(set) var state: PuzzleDeviceState = .disconnected
    func connect(sessionId: String) async {}
    func send(_ command: DeviceCommand) async throws {}
    func disconnect() async {}
}
final class FakePuzzleDeviceConnector: PuzzleDeviceConnector {
    var state: PuzzleDeviceState = .disconnected
    func connect(sessionId: String) async { state = .connecting; state = .connected }
    func send(_ command: DeviceCommand) async throws {}
    func disconnect() async { state = .disconnected }
}
protocol PuzzleOutput { func puzzleDidComplete(id: String) async }
struct NoOpPuzzleOutput: PuzzleOutput { func puzzleDidComplete(id: String) async {} }

File: AOS/app/src/main/java/com/example/escapephone/core/services/PlatformServices.kt
package com.example.escapephone.core.services

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import kotlinx.coroutines.delay

enum class HapticEvent { success, error, selection, digitDiscovered }
interface HapticProvider { fun play(event: HapticEvent) }
class PlatformHapticProvider(context: Context) : HapticProvider {
    private val vibrator: Vibrator = if (Build.VERSION.SDK_INT >= 31) context.getSystemService(VibratorManager::class.java).defaultVibrator else @Suppress("DEPRECATION") (context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator)
    override fun play(event: HapticEvent) { val duration = if (event == HapticEvent.error) 70L else 30L; vibrator.vibrate(VibrationEffect.createOneShot(duration, VibrationEffect.DEFAULT_AMPLITUDE)) }
}
object NoOpHapticProvider : HapticProvider { override fun play(event: HapticEvent) = Unit }

sealed interface AdResult { data object rewarded : AdResult; data object skipped : AdResult; data object unavailable : AdResult; data class failed(val message: String) : AdResult }
interface AdGateway { suspend fun showRewardedAd(): AdResult }
object NoOpAdGateway : AdGateway { override suspend fun showRewardedAd() = AdResult.unavailable }
class FakeAdGateway(var result: AdResult = AdResult.rewarded) : AdGateway { override suspend fun showRewardedAd(): AdResult { delay(300); return result } }

enum class DeviceCommand { ping, arm, unlock, reset }
sealed interface PuzzleDeviceState { data object disconnected : PuzzleDeviceState; data object connecting : PuzzleDeviceState; data object connected : PuzzleDeviceState; data class error(val message: String) : PuzzleDeviceState }
interface PuzzleDeviceConnector { val state: PuzzleDeviceState; suspend fun connect(sessionId: String); suspend fun send(command: DeviceCommand); suspend fun disconnect() }
class NoOpPuzzleDeviceConnector : PuzzleDeviceConnector { override val state = PuzzleDeviceState.disconnected; override suspend fun connect(sessionId: String) = Unit; override suspend fun send(command: DeviceCommand) = Unit; override suspend fun disconnect() = Unit }
class FakePuzzleDeviceConnector : PuzzleDeviceConnector { override var state: PuzzleDeviceState = PuzzleDeviceState.disconnected; override suspend fun connect(sessionId: String) { state = PuzzleDeviceState.connecting; state = PuzzleDeviceState.connected }; override suspend fun send(command: DeviceCommand) = Unit; override suspend fun disconnect() { state = PuzzleDeviceState.disconnected } }
interface PuzzleOutput { suspend fun puzzleDidComplete(id: String) }
object NoOpPuzzleOutput : PuzzleOutput { override suspend fun puzzleDidComplete(id: String) = Unit }
